tests/test_color.py

```diff
@@ -0,0 +1,315 @@
+        ((60, 0.75, 0.25), (0x70, 0x70, 0x10)),
+        ((120, 0.75, 0.25), (0x10, 0x70, 0x10)),
+        ((180, 0.75, 0.25), (0x10, 0x70, 0x70)),
+        ((240, 0.75, 0.25), (0x10, 0x10, 0x70)),
+        ((300, 0.75, 0.25), (0x70, 0x10, 0x70)),
+        ((360, 0.75, 0.25), (0x70, 0x10, 0x10)),
+    ],
+)
+def test_hsl(value, expected):
+    assert_equal_color(hsl(*value), rgb(*expected))
+
+
+@pytest.mark.parametrize(
+    "value, expected",
+    [
+        ((60, 0.0, 0.0, 0.3), (0x00, 0x00, 0x00, 0.3)),
+        ((60, 0.0, 1.0, 0.3), (0xFF, 0xFF, 0xFF, 0.3)),
+        ((60, 1.0, 0.5, 0.3), (0xFF, 0xFF, 0x00, 0.3)),
+        ((60, 0.25, 0.25, 0.3), (0x50, 0x50, 0x30, 0.3)),
+        ((60, 0.25, 0.75, 0.3), (0xCF, 0xCF, 0xAF, 0.3)),
+        ((60, 0.75, 0.75, 0.3), (0xEF, 0xEF, 0x8F, 0.3)),
+        ((60, 0.75, 0.25, 0.3), (0x70, 0x70, 0x10, 0.3)),
+    ],
+)
+def test_hsl_alpha(value, expected):
+    assert_equal_color(hsla(*value), rgba(*expected))
+
+
+@pytest.mark.parametrize(
+    "constructor, value, name, min, max, actual",
+    [
+        (rgb, (-1, 120, 10), "red", 0, 255, -1),
+        (rgb, (256, 120, 10), "red", 0, 255, 256),
+        (rgb, (120, -1, 10), "green", 0, 255, -1),
+        (rgb, (120, 256, 10), "green", 0, 255, 256),
+        (rgb, (120, 10, -1), "blue", 0, 255, -1),
+        (rgb, (120, 10, 256), "blue", 0, 255, 256),
+        (rgba, (-1, 120, 10, 0.5), "red", 0, 255, -1),
+        (rgba, (256, 120, 10, 0.5), "red", 0, 255, 256),
+        (rgba, (120, -1, 10, 0.5), "green", 0, 255, -1),
+        (rgba, (120, 256, 10, 0.5), "green", 0, 255, 256),
+        (rgba, (120, 10, -1, 0.5), "blue", 0, 255, -1),
+        (rgba, (120, 10, 256, 0.5), "blue", 0, 255, 256),
+        (rgba, (120, 10, 60, -0.5), "alpha", 0, 1, -0.5),
+        (rgba, (120, 10, 60, 1.1), "alpha", 0, 1, 1.1),
+        (hsl, (-1, 0.5, 0.8), "hue", 0, 360, -1),
+        (hsl, (361, 0.5, 0.8), "hue", 0, 360, 361),
+        (hsl, (120, -0.1, 0.8), "saturation", 0, 1, -0.1),
+        (hsl, (120, 1.1, 0.8), "saturation", 0, 1, 1.1),
+        (hsl, (120, 0.8, -0.1), "lightness", 0, 1, -0.1),
+        (hsl, (120, 0.8, 1.1), "lightness", 0, 1, 1.1),
+        (hsla, (-1, 0.5, 0.8, 0.5), "hue", 0, 360, -1),
+        (hsla, (361, 0.5, 0.8, 0.5), "hue", 0, 360, 361),
+        (hsla, (120, -0.1, 0.8, 0.5), "saturation", 0, 1, -0.1),
+        (hsla, (120, 1.1, 0.8, 0.5), "saturation", 0, 1, 1.1),
+        (hsla, (120, 0.8, -0.1, 0.5), "lightness", 0, 1, -0.1),
+        (hsla, (120, 0.8, 1.1, 0.5), "lightness", 0, 1, 1.1),
+        (hsla, (120, 0.8, 0.5, -0.1), "alpha", 0, 1, -0.1),
+        (hsla, (120, 0.8, 0.5, 1.1), "alpha", 0, 1, 1.1),
+    ],
+)
+def test_invalid_color_constructor(constructor, value, name, min, max, actual):
+    with pytest.raises(
+        ValueError,
+        match=rf"^{name} value should be between {min}-{max}\. Got {actual}$",
+    ):
+        constructor(*value)
+
+
+###############
+# Color parsing
+###############
```

Comment:
I agree it makes sense to merge the exception tests in here; but the parsing tests should probably be broken out, for the same reason as the font tests.